File: services/server/project/qld/ORETrade/ir/oiswap.py
"""
OIS (Overnight Index Swap) Trade Implementation using ORE.

This module provides OIS pricing using the Open Risk Engine (ORE).
"""

import pandas as pd
import project.qld.engine.parser.date as dateParser

from ..ore_base_trade import OREBaseTrade


class OISwap(OREBaseTrade):
    """
    Overnight Index Swap (OIS) trade priced using ORE.

    An OIS is an interest rate swap where the floating leg is tied
    to an overnight rate index (e.g., EONIA, SOFR).
    """

    __name__ = "OvernightIndexSwap"

    def __init__(self, evaluation_date, currency, notional, payer, start_date, end_date,
                 tenor, calendar, convention, index, spread, is_in_arrears, fixing_days,
                 day_counter, payment_convention):
        """
        Initialize OIS trade.

        Args:
            evaluation_date: Valuation date
            currency: Currency (e.g., 'EUR', 'USD')
            notional: Notional amount
            payer: True if paying floating, False if receiving floating
            start_date: Start date of the swap
            end_date: End date of the swap
            tenor: Payment frequency (e.g., '3M', '6M')
            calendar: Business day calendar (e.g., 'TARGET', 'US')
            convention: Business day convention (e.g., 'MF', 'Following')
            index: Floating rate index (e.g., 'EUR-EONIA', 'USD-SOFR')
            spread: Spread over index (in decimal, e.g., 0.001 for 10bp)
            is_in_arrears: Whether fixing is in arrears
            fixing_days: Number of fixing days
            day_counter: Day count convention (e.g., 'A360', 'ACT/360')
            payment_convention: Payment convention
        """
        super().__init__(evaluation_date)

        self.currency = currency
        self.notional = notional
        self.payer = payer
        self.start_date = start_date
        self.end_date = end_date
        self.tenor = tenor
        self.calendar = calendar
        self.convention = convention
        self.index = index
        self.spread = spread
        self.is_in_arrears = is_in_arrears
        self.fixing_days = fixing_days
        self.day_counter = day_counter
        self.payment_convention = payment_convention

        self._parse_input()
        self._validate_input()

    def _parse_input(self):
        """Parse and convert string inputs to appropriate types."""
        self.evaluation_date = dateParser.parse_date(self.evaluation_date)
        self.currency = self.currency.upper()
        self.notional = float(self.notional)
        self.start_date = dateParser.parse_date(self.start_date)
        self.end_date = dateParser.parse_date(self.end_date)
        self.spread = float(self.spread)
        self.fixing_days = int(self.fixing_days)

        # Parse boolean payer flag
        if isinstance(self.payer, str):
            self.payer = self.payer.lower() in ('true', 'yes', '1')
        else:
            self.payer = bool(self.payer)

        # Parse boolean is_in_arrears
        if isinstance(self.is_in_arrears, str):
            self.is_in_arrears = self.is_in_arrears.lower() in ('true', 'yes', '1')
        else:
            self.is_in_arrears = bool(self.is_in_arrears)

    def _validate_input(self):
        """Validate trade parameters."""
        if self.notional <= 0:
            raise ValueError("Notional must be positive")

        if self.start_date >= self.end_date:
            raise ValueError("End date must be after start date")

        if self.start_date < self.evaluation_date:
            raise ValueError("Start date cannot be before evaluation date")

        if self.fixing_days < 0:
            raise ValueError("Fixing days must be non-negative")

    def _build_trade_dataframe(self):
        """
        Build pandas DataFrame with OIS trade data for ORE.

        Returns:
            tuple: ('Swap', pd.DataFrame with trade parameters)

        Note: OIS is represented as a Swap with only a floating leg in ORE.
        """
        trade_df = pd.DataFrame({
            'TradeId': [self.__name__],
            'TradeType': [self.__name__],
            'CounterParty': ['CPTY_A'],
            'Currency': [self.currency],
            'Notional': [self.notional],
            'Payer': [str(self.payer).lower()],
            'StartDate': [dateParser.ql2pydate(self.start_date).strftime('%Y-%m-%d')],
            'EndDate': [dateParser.ql2pydate(self.end_date).strftime('%Y-%m-%d')],
            'Tenor': [self.tenor],
            'Calendar': [self.calendar],
            'Convention': [self.convention],
            'FloatingIndex': [self.index],
            'FloatingSpread': [self.spread],
            'IsInArrears': [str(self.is_in_arrears).lower()],
            'FixingDays': [self.fixing_days],
            'DayCounter': [self.day_counter],
            'PaymentConvention': [self.payment_convention],
            # For OIS, we only have a floating leg
            'LegType': ['Floating'],
        })

        return (self.__name__, trade_df)

    def _filter_epe_dates(self, exposure_df):
        """
        Filter exposure DataFrame to include only relevant dates.

        Args:
            exposure_df (pd.DataFrame): DataFrame with exposure data including 'Date' column.
        Returns:
            pd.DataFrame: Filtered DataFrame.
        """
        exposure_df['Date'] = pd.to_datetime(exposure_df['Date'])
        start_date = dateParser.ql2npdate(self.start_date)
        end_date = dateParser.ql2npdate(self.end_date)
        filtered_df = exposure_df[(exposure_df['Date'] >= start_date) & (exposure_df['Date'] <= end_date)]
        return filtered_df

File: services/client/src/forms/ir/oiswap.js
import { sensitivityGroup } from '../shared/sensitivity-fields'

export const oiswapForm = {
  name: 'Overnight Index Swap',
  category: 'IR',
  api: 'calculateIROISwap',
  groups: [
    {
      title: 'Trade Details',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'evaluation_date',
          type: 'date',
          label: 'Evaluation Date',
          required: true,
          defaultValue: new Date('2016-02-05'),
          helpText: 'The date at which the swap will be valued'
        },
        {
          key: 'currency',
          type: 'dropdown',
          label: 'Currency',
          required: true,
          defaultValue: 'EUR',
          options: [
            { label: 'USD - US Dollar', value: 'USD' },
            { label: 'EUR - Euro', value: 'EUR' },
            { label: 'GBP - British Pound', value: 'GBP' },
            { label: 'JPY - Japanese Yen', value: 'JPY' },
            { label: 'AUD - Australian Dollar', value: 'AUD' },
            { label: 'CAD - Canadian Dollar', value: 'CAD' },
            { label: 'CHF - Swiss Franc', value: 'CHF' }
          ],
          filter: true,
          helpText: 'The currency in which the swap is denominated'
        },
        {
          key: 'notional',
          type: 'float',
          label: 'Notional Amount',
          required: true,
          defaultValue: 100000000,
          step: 1000,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          helpText: 'The notional amount of the swap',
          validators: [
            (value) => {
              if (value <= 0) {
                return 'Notional amount must be greater than 0'
              }
              return true
            }
          ]
        },
        {
          key: 'payer',
          type: 'boolean',
          label: 'Payer',
          required: true,
          defaultValue: false,
          helpText: 'True if paying floating rate, false if receiving floating rate'
        },
        {
          key: 'start_date',
          type: 'date',
          label: 'Start Date',
          required: true,
          defaultValue: new Date('2023-01-31'),
          helpText: 'The effective date when the swap begins',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.evaluation_date) return true
              if (value < formData.evaluation_date) {
                return 'Start date cannot be before evaluation date'
              }
              return true
            }
          ]
        },
        {
          key: 'end_date',
          type: 'date',
          label: 'Maturity Date',
          required: true,
          defaultValue: new Date('2028-02-01'),
          helpText: 'The date when the swap matures',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.start_date) return true
              if (value <= formData.start_date) {
                return 'Maturity date must be after start date'
              }
              return true
            }
          ]
        },
        {
          key: 'tenor',
          type: 'dropdown',
          label: 'Payment Frequency',
          required: true,
          defaultValue: '3M',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently payments are made'
        },
        {
          key: 'calendar',
          type: 'dropdown',
          label: 'Business Day Calendar',
          required: true,
          defaultValue: 'US',
          options: [
            { label: 'US - United States', value: 'US' },
            { label: 'TARGET - European Central Bank', value: 'TARGET' },
            { label: 'UK - United Kingdom', value: 'UK' },
            { label: 'JP - Japan', value: 'JP' },
            { label: 'AU - Australia', value: 'AU' },
            { label: 'CA - Canada', value: 'CA' },
            { label: 'CH - Switzerland', value: 'CH' }
          ],
          filter: true,
          helpText: 'The business day calendar used for date adjustments'
        },
        {
          key: 'convention',
          type: 'dropdown',
          label: 'Business Day Convention',
          required: true,
          defaultValue: 'MF',
          options: [
            { label: 'Following', value: 'Following' },
            { label: 'Modified Following', value: 'MF' },
            { label: 'Preceding', value: 'Preceding' },
            { label: 'Modified Preceding', value: 'ModifiedPreceding' }
          ],
          filter: true,
          helpText: 'The convention used to adjust dates that fall on non-business days'
        }
      ]
    },
    {
      title: 'Floating Leg (Overnight Index)',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'index',
          type: 'dropdown',
          label: 'Overnight Index',
          required: true,
          defaultValue: 'EUR-EONIA',
          options: [
            { label: 'EUR-EONIA - Euro OverNight Index Average', value: 'EUR-EONIA' },
            { label: 'USD-SOFR - Secured Overnight Financing Rate', value: 'USD-SOFR' },
            { label: 'GBP-SONIA - Sterling Overnight Index Average', value: 'GBP-SONIA' },
            { label: 'JPY-TONAR - Tokyo Overnight Average Rate', value: 'JPY-TONAR' },
            { label: 'AUD-AONIA - Australian Overnight Index Average', value: 'AUD-AONIA' },
            { label: 'CAD-CORRA - Canadian Overnight Repo Rate Average', value: 'CAD-CORRA' },
            { label: 'CHF-SARON - Swiss Average Rate Overnight', value: 'CHF-SARON' }
          ],
          filter: true,
          helpText: 'The overnight index used for the floating rate'
        },
        {
          key: 'spread',
          type: 'float',
          label: 'Spread',
          required: true,
          defaultValue: 0.000122,
          step: 0.000001,
          minFractionDigits: 6,
          maxFractionDigits: 8,
          helpText: 'Spread over the overnight index in decimal form (e.g., 0.000122 for 1.22bp)'
        },
        {
          key: 'is_in_arrears',
          type: 'boolean',
          label: 'Is In Arrears',
          required: true,
          defaultValue: false,
          helpText: 'Whether the fixing is in arrears (fixed at end of period)'
        },
        {
          key: 'fixing_days',
          type: 'number',
          label: 'Fixing Days',
          required: true,
          defaultValue: 2,
          min: 0,
          max: 10,
          helpText: 'Number of business days between rate fixing and payment period start',
          validators: [
            (value) => {
              if (value < 0) {
                return 'Fixing days cannot be negative'
              }
              if (value > 10) {
                return 'Fixing days cannot exceed 10'
              }
              return true
            }
          ]
        },
        {
          key: 'day_counter',
          type: 'dropdown',
          label: 'Day Count Convention',
          required: true,
          defaultValue: 'A360',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText: 'The convention used to calculate the fraction of a year for payments'
        },
        {
          key: 'payment_convention',
          type: 'dropdown',
          label: 'Payment Convention',
          required: true,
          defaultValue: 'MF',
          options: [
            { label: 'Following', value: 'Following' },
            { label: 'Modified Following', value: 'MF' },
            { label: 'Preceding', value: 'Preceding' },
            { label: 'Modified Preceding', value: 'ModifiedPreceding' }
          ],
          filter: true,
          helpText: 'Business day convention for payment dates'
        }
      ]
    },
    sensitivityGroup
  ]
}
